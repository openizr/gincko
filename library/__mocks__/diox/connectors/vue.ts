/**
 * Copyright (c) Openizr. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

/**
 * diox/connectors/vue mock.
 */
const connect = vi.fn(() => vi.fn(() => ({
  steps: [{ id: 'start' }, { id: 'end' }],
  loading: process.env.LOADING === 'true',
})));

export default connect;
