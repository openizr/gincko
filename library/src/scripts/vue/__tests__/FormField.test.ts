/**
 * Copyright (c) Openizr. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @vitest-environment jsdom
 */

import FormField from 'scripts/vue/FormField.vue';
import { render, waitFor } from '@testing-library/vue';

vi.mock('biuty/vue');
vi.mock('scripts/core/Engine');
vi.mock('scripts/vue/MessageField.vue');
vi.mock('scripts/vue/NestedFields.vue');

const flushPromise = vi.fn(() => new Promise<void>((resolve) => { setTimeout(resolve, 50); }));

describe('vue/FormField', () => {
  const isActive = true;
  const customComponents = {};
  const userInputs = { test: 'ok' };
  const variables = { var: 'value' };
  const i18n = vi.fn(() => 'TRANSLATED LABEL');
  const onUserAction = vi.fn() as unknown as OnUserAction;

  beforeEach(() => {
    vi.clearAllMocks();
    delete process.env.IS_DATE;
  });

  test('renders correctly - Unknown component', async () => {
    const { container } = render(FormField, {
      props: {
        isActive,
        i18n,
        path: 'path.0.to.field',
        variables,
        userInputs,
        onUserAction,
        customComponents,
        field: {
          id: 'field',
          status: 'success',
          component: 'Unknown',
          componentProps: {},
        },
      },
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - Message with label and helper', async () => {
    const { container } = render(FormField, {
      props: {
        isActive,
        i18n,
        path: 'path.0.to.field',
        variables,
        userInputs,
        onUserAction,
        customComponents,
        field: {
          id: 'field',
          label: 'Test',
          status: 'success',
          component: 'Message',
          componentProps: { helper: 'Helper' },
        },
      },
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - Message with label and message', async () => {
    const { container } = render(FormField, {
      props: {
        isActive,
        i18n,
        path: 'path.0.to.field',
        variables,
        userInputs,
        onUserAction,
        customComponents,
        field: {
          id: 'field',
          label: 'Test',
          message: 'Success',
          status: 'success',
          component: 'Message',
          componentProps: { helper: 'Helper' },
        },
      },
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - Button', async () => {
    const { container } = render(FormField, {
      props: {
        isActive,
        i18n,
        path: 'path.0.to.field',
        variables,
        userInputs,
        onUserAction,
        customComponents,
        field: {
          id: 'field',
          status: 'success',
          component: 'Button',
          componentProps: {},
        },
      },
    });
    await waitFor(flushPromise);
    expect(container.firstChild).toMatchSnapshot();
    expect(onUserAction).toHaveBeenCalledTimes(1);
    expect(onUserAction).toHaveBeenCalledWith('input', 'path.0.to.field', true);
  });

  test('renders correctly - Link', async () => {
    const { container } = render(FormField, {
      props: {
        isActive,
        i18n,
        path: 'path.0.to.field',
        variables,
        userInputs,
        onUserAction,
        customComponents,
        field: {
          id: 'field',
          status: 'success',
          component: 'Link',
          componentProps: {},
        },
      },
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - Options', async () => {
    const { container } = render(FormField, {
      props: {
        isActive,
        i18n,
        path: 'path.0.to.field',
        variables,
        userInputs,
        onUserAction,
        customComponents,
        field: {
          id: 'field',
          status: 'success',
          component: 'Options',
          componentProps: {
            onFocus: (): null => null,
            options: [{ label: 'test', value: 'option1' }, { value: 'option2' }],
          },
        },
      },
    });
    await waitFor(flushPromise);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - DatePicker', async () => {
    process.env.IS_DATE = 'true';
    const { container, rerender } = render(FormField, {
      props: {
        isActive,
        i18n,
        path: 'path.0.to.field',
        variables,
        userInputs,
        onUserAction,
        customComponents,
        field: {
          id: 'field',
          status: 'success',
          component: 'DatePicker',
          componentProps: {},
          value: null,
        },
      },
    });
    await waitFor(flushPromise);
    expect(container.firstChild).toMatchSnapshot();
    await rerender({
      isActive,
      i18n,
      path: 'path.0.to.field',
      variables,
      userInputs,
      onUserAction,
      customComponents,
      field: {
        id: 'field',
        status: 'success',
        component: 'DatePicker',
        value: new Date(1657192371401),
        componentProps: { placeholder: 'placeholder', debounceTimeout: 10 },
      },
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - Textfield', async () => {
    const { container, rerender } = render(FormField, {
      props: {
        isActive,
        i18n,
        path: 'path.0.to.field',
        variables,
        userInputs,
        onUserAction,
        customComponents,
        field: {
          id: 'field',
          value: null,
          status: 'success',
          component: 'Textfield',
          componentProps: {},
        },
      },
    });
    await waitFor(flushPromise);
    expect(container.firstChild).toMatchSnapshot();
    await rerender({
      isActive,
      i18n,
      path: 'path.0.to.field',
      variables,
      userInputs,
      onUserAction,
      customComponents,
      field: {
        id: 'field',
        value: 'test',
        status: 'success',
        component: 'Textfield',
        componentProps: { placeholder: 'placeholder', debounceTimeout: 10 },
      },
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - Textarea', async () => {
    const { container, rerender } = render(FormField, {
      props: {
        isActive,
        i18n,
        path: 'path.0.to.field',
        variables,
        userInputs,
        onUserAction,
        customComponents,
        field: {
          id: 'field',
          value: null,
          status: 'success',
          component: 'Textarea',
          componentProps: {},
        },
      },
    });
    await waitFor(flushPromise);
    expect(container.firstChild).toMatchSnapshot();
    await rerender({
      isActive,
      i18n,
      path: 'path.0.to.field',
      variables,
      userInputs,
      onUserAction,
      customComponents,
      field: {
        id: 'field',
        value: 'test',
        status: 'success',
        component: 'Textarea',
        componentProps: { placeholder: 'placeholder', debounceTimeout: 10 },
      },
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - FilePicker', async () => {
    const { container, rerender } = render(FormField, {
      props: {
        isActive,
        i18n,
        path: 'path.0.to.field',
        variables,
        userInputs,
        onUserAction,
        customComponents,
        field: {
          id: 'field',
          status: 'success',
          component: 'FilePicker',
          componentProps: {},
        },
      },
    });
    await waitFor(flushPromise);
    expect(container.firstChild).toMatchSnapshot();
    await rerender({
      isActive,
      i18n,
      path: 'path.0.to.field',
      variables,
      userInputs,
      onUserAction,
      customComponents,
      field: {
        id: 'field',
        status: 'success',
        component: 'FilePicker',
        componentProps: { placeholder: 'placeholder' },
      },
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - Array', async () => {
    const { container, rerender } = render(FormField, {
      props: {
        isActive,
        i18n,
        path: 'path.0.to.field',
        variables,
        userInputs,
        onUserAction,
        customComponents,
        field: {
          id: 'field',
          status: 'success',
          component: 'Array',
          componentProps: {},
          fields: [],
        },
      },
    });
    expect(container.firstChild).toMatchSnapshot();
    await rerender({
      isActive: false,
      i18n,
      path: 'path.0.to.field',
      variables,
      userInputs,
      onUserAction,
      customComponents,
      field: {
        id: 'field',
        status: 'success',
        component: 'Array',
        componentProps: {
          addButtonProps: { label: 'add' },
          removeButtonProps: { label: 'remove' },
        },
        fields: [],
      },
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - DynamicObject', async () => {
    const { container, rerender } = render(FormField, {
      props: {
        isActive,
        i18n,
        path: 'path.0.to.field',
        variables,
        userInputs,
        onUserAction,
        customComponents,
        field: {
          id: 'field',
          status: 'success',
          component: 'DynamicObject',
          componentProps: {},
          fields: [],
        },
      },
    });
    expect(container.firstChild).toMatchSnapshot();
    await rerender({
      isActive,
      i18n,
      path: 'path.0.to.field',
      variables,
      userInputs,
      onUserAction,
      customComponents,
      field: {
        id: 'field',
        status: 'success',
        component: 'DynamicObject',
        componentProps: {
          addTextfieldProps: { label: 'pattern', placeholder: 'placeholder' },
        },
        fields: [],
      },
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
