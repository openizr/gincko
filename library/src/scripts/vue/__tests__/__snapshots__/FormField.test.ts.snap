// Vitest Snapshot v1

exports[`vue/FormField > renders correctly - Array 1`] = `
<div
  addbuttonprops="[object Object]"
  addtextfieldprops="[object Object]"
  customcomponents="[object Object]"
  fields=""
  id="path__0__to__field"
  isactive="true"
  modifiers="success "
  removebuttonprops="[object Object]"
  t="function(...o) {
    if (e.called = !0, e.callCount++, e.calls.push(o), e.next) {
      let [s, l] = e.next;
      if (e.results.push(e.next), e.next = null, s === \\"ok\\")
        return l;
      throw l;
    }
    let r, n = \\"ok\\";
    if (e.impl)
      try {
        r = e.impl.apply(this, o), n = \\"ok\\";
      } catch (s) {
        throw r = s, n = \\"error\\", e.results.push([n, s]), s;
      }
    let c = [n, r];
    if (v(r)) {
      let s = r.then((l) => c[1] = l).catch((l) => {
        throw c[0] = \\"error\\", c[1] = l, l;
      });
      Object.assign(s, r), r = s;
    }
    return e.results.push(c), r;
  }"
  type="array"
  userinputs="[object Object]"
  variables="[object Object]"
>
  {"path":"path.0.to.field"}
</div>
`;

exports[`vue/FormField > renders correctly - Array 2`] = `
<div
  addbuttonprops="[object Object]"
  addtextfieldprops="[object Object]"
  customcomponents="[object Object]"
  fields=""
  id="path__0__to__field"
  isactive="true"
  modifiers="success "
  removebuttonprops="[object Object]"
  t="function(...o) {
    if (e.called = !0, e.callCount++, e.calls.push(o), e.next) {
      let [s, l] = e.next;
      if (e.results.push(e.next), e.next = null, s === \\"ok\\")
        return l;
      throw l;
    }
    let r, n = \\"ok\\";
    if (e.impl)
      try {
        r = e.impl.apply(this, o), n = \\"ok\\";
      } catch (s) {
        throw r = s, n = \\"error\\", e.results.push([n, s]), s;
      }
    let c = [n, r];
    if (v(r)) {
      let s = r.then((l) => c[1] = l).catch((l) => {
        throw c[0] = \\"error\\", c[1] = l, l;
      });
      Object.assign(s, r), r = s;
    }
    return e.results.push(c), r;
  }"
  type="array"
  userinputs="[object Object]"
  variables="[object Object]"
>
  {"path":"path.0.to.field"}
</div>
`;

exports[`vue/FormField > renders correctly - Button 1`] = `
<div
  id="path__0__to__field"
  modifiers="success "
>
  {}
</div>
`;

exports[`vue/FormField > renders correctly - DatePicker 1`] = `
<div
  debouncetimeout="100"
  id="path__0__to__field"
  maxlength="10"
  modifiers="success "
  name="path.0.to.field"
>
  {}
</div>
`;

exports[`vue/FormField > renders correctly - DatePicker 2`] = `
<div
  debouncetimeout="100"
  id="path__0__to__field"
  maxlength="10"
  modifiers="success "
  name="path.0.to.field"
>
  {}
</div>
`;

exports[`vue/FormField > renders correctly - DynamicObject 1`] = `
<div
  addbuttonprops="[object Object]"
  addtextfieldprops="[object Object]"
  customcomponents="[object Object]"
  fields=""
  id="path__0__to__field"
  isactive="true"
  modifiers="success "
  removebuttonprops="[object Object]"
  t="function(...o) {
    if (e.called = !0, e.callCount++, e.calls.push(o), e.next) {
      let [s, l] = e.next;
      if (e.results.push(e.next), e.next = null, s === \\"ok\\")
        return l;
      throw l;
    }
    let r, n = \\"ok\\";
    if (e.impl)
      try {
        r = e.impl.apply(this, o), n = \\"ok\\";
      } catch (s) {
        throw r = s, n = \\"error\\", e.results.push([n, s]), s;
      }
    let c = [n, r];
    if (v(r)) {
      let s = r.then((l) => c[1] = l).catch((l) => {
        throw c[0] = \\"error\\", c[1] = l, l;
      });
      Object.assign(s, r), r = s;
    }
    return e.results.push(c), r;
  }"
  type="dynamicObject"
  userinputs="[object Object]"
  variables="[object Object]"
>
  {"path":"path.0.to.field"}
</div>
`;

exports[`vue/FormField > renders correctly - DynamicObject 2`] = `
<div
  addbuttonprops="[object Object]"
  addtextfieldprops="[object Object]"
  customcomponents="[object Object]"
  fields=""
  id="path__0__to__field"
  isactive="true"
  modifiers="success "
  removebuttonprops="[object Object]"
  t="function(...o) {
    if (e.called = !0, e.callCount++, e.calls.push(o), e.next) {
      let [s, l] = e.next;
      if (e.results.push(e.next), e.next = null, s === \\"ok\\")
        return l;
      throw l;
    }
    let r, n = \\"ok\\";
    if (e.impl)
      try {
        r = e.impl.apply(this, o), n = \\"ok\\";
      } catch (s) {
        throw r = s, n = \\"error\\", e.results.push([n, s]), s;
      }
    let c = [n, r];
    if (v(r)) {
      let s = r.then((l) => c[1] = l).catch((l) => {
        throw c[0] = \\"error\\", c[1] = l, l;
      });
      Object.assign(s, r), r = s;
    }
    return e.results.push(c), r;
  }"
  type="dynamicObject"
  userinputs="[object Object]"
  variables="[object Object]"
>
  {"path":"path.0.to.field"}
</div>
`;

exports[`vue/FormField > renders correctly - FilePicker 1`] = `
<div
  id="path__0__to__field"
  modifiers="success "
  name="path.0.to.field"
>
  {}
</div>
`;

exports[`vue/FormField > renders correctly - FilePicker 2`] = `
<div
  id="path__0__to__field"
  modifiers="success "
  name="path.0.to.field"
>
  {}
</div>
`;

exports[`vue/FormField > renders correctly - Link 1`] = `
<div
  id="path__0__to__field"
  label=""
  modifiers="success "
>
  {}
</div>
`;

exports[`vue/FormField > renders correctly - Message with label and helper 1`] = `
<div
  id="field"
>
  {"id":"path__0__to__field","label":"TRANSLATED LABEL","modifiers":"success "}
</div>
`;

exports[`vue/FormField > renders correctly - Message with label and message 1`] = `
<div
  id="field"
>
  {"id":"path__0__to__field","label":"TRANSLATED LABEL","modifiers":"success "}
</div>
`;

exports[`vue/FormField > renders correctly - Options 1`] = `
<div
  id="path__0__to__field"
  modifiers="success "
  name="path.0.to.field"
  options="[object Object],[object Object]"
>
  {}
</div>
`;

exports[`vue/FormField > renders correctly - Textarea 1`] = `
<div
  debouncetimeout="100"
  id="path__0__to__field"
  modifiers="success "
  name="path.0.to.field"
>
  {}
</div>
`;

exports[`vue/FormField > renders correctly - Textarea 2`] = `
<div
  debouncetimeout="100"
  id="path__0__to__field"
  modifiers="success "
  name="path.0.to.field"
>
  {}
</div>
`;

exports[`vue/FormField > renders correctly - Textfield 1`] = `
<div
  debouncetimeout="100"
  id="path__0__to__field"
  modifiers="success "
  name="path.0.to.field"
>
  {}
</div>
`;

exports[`vue/FormField > renders correctly - Textfield 2`] = `
<div
  debouncetimeout="100"
  id="path__0__to__field"
  modifiers="success "
  name="path.0.to.field"
>
  {}
</div>
`;
